src/hid-asus.c

```diff
@@ -0,0 +1,320 @@
+MODULE_AUTHOR("Victor Vlasenko <victor.vlasenko@sysgears.com>");
+MODULE_AUTHOR("Frederik Wenigwieser <frederik.wenigwieser@gmail.com>");
+MODULE_DESCRIPTION("Asus HID Keyboard and TouchPad");
+
+#define FEATURE_REPORT_ID 0x0d
+#define INPUT_REPORT_ID 0x5d
+
+#define INPUT_REPORT_SIZE 28
+
+#define MAX_CONTACTS 5
+
+#define MAX_X 0x0aea
+#define MAX_Y 0x06de
+#define MAX_TOUCH_MAJOR 8
+#define MAX_PRESSURE 0x80
+
+
+#define REPORT_ID_OFFSET 0
+
+#define CONTACT_DOWN_OFFSET 1
+#define CONTACT_DOWN_MASK 0x08
+
+#define BTN_LEFT_OFFSET 1
+#define BTN_LEFT_MASK 0x01
+
+#define CONTACT_DATA_OFFSET 2
+#define CONTACT_DATA_SIZE 5
+
+#define CONTACT_TOOL_TYPE_OFFSET 3
+#define CONTACT_TOOL_TYPE_MASK 0x80
+
+#define CONTACT_X_MSB_OFFSET 0
+#define CONTACT_X_MSB_BIT_SHIFT 4
+#define CONTACT_X_LSB_OFFSET 1
+
+#define CONTACT_Y_MSB_OFFSET 0
+#define CONTACT_Y_MSB_MASK 0x0f
+#define CONTACT_Y_LSB_OFFSET 2
+
+#define CONTACT_TOUCH_MAJOR_OFFSET 3
+#define CONTACT_TOUCH_MAJOR_BIT_SHIFT 4
+#define CONTACT_TOUCH_MAJOR_MASK 0x07
+
+#define CONTACT_PRESSURE_OFFSET 4
+#define CONTACT_PRESSURE_MASK 0x7f
+
+#define BYTE_BIT_SHIFT 8
+#define TRKID_SGN       ((TRKID_MAX + 1) >> 1)
+
+static void asus_report_contact_down(struct input_dev *input,
+		int toolType, u8 *data)
+{
+	int x, y, touch_major, pressure;
+
+	x = data[CONTACT_X_MSB_OFFSET] >> CONTACT_X_MSB_BIT_SHIFT;
+	x = (x << BYTE_BIT_SHIFT) | data[CONTACT_X_LSB_OFFSET];
+
+	y = (data[CONTACT_Y_MSB_OFFSET] & CONTACT_Y_MSB_MASK);
+	y = MAX_Y - ((y << BYTE_BIT_SHIFT) | data[CONTACT_Y_LSB_OFFSET]);
+
+	if (toolType == MT_TOOL_PALM) {
+		touch_major = MAX_TOUCH_MAJOR;
+		pressure = MAX_PRESSURE;
+	} else {
+		touch_major = data[CONTACT_TOUCH_MAJOR_OFFSET];
+		touch_major >>= CONTACT_TOUCH_MAJOR_BIT_SHIFT;
+		touch_major &= CONTACT_TOUCH_MAJOR_MASK;
+		pressure = data[CONTACT_PRESSURE_OFFSET] & CONTACT_PRESSURE_MASK;
+	}
+
+	input_report_abs(input, ABS_MT_POSITION_X, x);
+	input_report_abs(input, ABS_MT_POSITION_Y, y);
+	input_report_abs(input, ABS_MT_TOUCH_MAJOR, touch_major);
+	input_report_abs(input, ABS_MT_PRESSURE, pressure);
+}
+
+/* Required for Synaptics Palm Detection */
+static void asus_report_tool_width(struct input_dev *input)
+{
+	struct input_mt *mt = input->mt;
+	struct input_mt_slot *oldest;
+	int oldid, count, i;
+
+	oldest = NULL;
+	oldid = mt->trkid;
+	count = 0;
+
+	for (i = 0; i < mt->num_slots; ++i) {
+		struct input_mt_slot *ps = &mt->slots[i];
+		int id = input_mt_get_value(ps, ABS_MT_TRACKING_ID);
+
+		if (id < 0)
+			continue;
+		if ((id - oldid) & TRKID_SGN) {
+			oldest = ps;
+			oldid = id;
+		}
+		count++;
+	}
+
+	if (oldest) {
+		input_report_abs(input, ABS_TOOL_WIDTH,
+			input_mt_get_value(oldest, ABS_MT_TOUCH_MAJOR));
+	}
+}
+
+static void asus_report_input(struct input_dev *input, u8 *data)
+{
+	int i;
+	u8 *contactData = data + CONTACT_DATA_OFFSET;
+
+	for (i = 0; i < MAX_CONTACTS; i++) {
+		bool down = data[CONTACT_DOWN_OFFSET] &
+				(CONTACT_DOWN_MASK << i);
```

Comment:
I'd force the value to be 0|1 by using a double negation ``!!``

Also, it might be easier to read if you use ``BIT(i+7)`` instead of ``(CONTACT_DOWN_MASK << i)``